File: src/python/problems/longest_valid_parentheses_test.py
# https://leetcode.com/problems/longest-valid-parentheses

from heapq import heappop, heappush
from core.problem_base import *

class LongestValidParentheses(ProblemBase):
    def Solution(self, s: str) -> int:
        stack = [-1]
        result = 0
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
                continue

            stack.pop()
            if not stack:
                stack.append(i)
            else:
                result = max(result, i - stack[-1])

        return result

if __name__ == '__main__':
    TestGen(LongestValidParentheses) \
        .Add(lambda tc: tc.Param("s", "(()").Result(2)) \
        .Add(lambda tc: tc.Param("s", ")()())").Result(4)) \
        .Add(lambda tc: tc.Param("s", "").Result(0)) \
        .Add(lambda tc: tc.Param("s", "()(())").Result(6)) \
        .Run()

File: src/go/problems/longest_valid_parentheses_test.go
// https://leetcode.com/problems/longest-valid-parentheses

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type LongestValidParentheses struct{}

func TestLongestValidParentheses(t *testing.T) {
	gen := core.TestSuite[LongestValidParentheses]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("(()").Result(2)
	}).Add(func(tc *core.TestCase) {
		tc.Param(")()())").Result(6)
	}).Add(func(tc *core.TestCase) {
		tc.Param("").Result(0)
	}).Add(func(tc *core.TestCase) {
		tc.Param("()(())").Result(6)
	}).Run(t)
}

func (LongestValidParentheses) Solution(s string) int {
    stack := []int{-1}
	max := 0
	length := len(s)
	for i := 0; i < length; i++ {
		if s[i] == '(' {
			stack = append(stack, i)
			continue
		}

		stack = stack[:len(stack) - 1]
		if len(stack) == 0 {
			stack = append(stack, i)
		} else {
			len := i - stack[len(stack) - 1]
			if max < len {
				max = len
			}
		}
	}

	return max
}
